README
This is an LV2 port of the MCP and VCO LADSPA plugins by Fons Adriaensen.

It features more precise plugin descriptions (in particular, CV ports are
distinct from audio ports, and unit metadata is available), but otherwise
remains faithful to the originals.  The code is identical to the original
LADSPA versions except for some type changes which should not affect output.
These plugins should be sonically indestinguishable from the originals.

Documentation for each plugin is included in the plugin's data file, and should
be available as online help in the host interface.

NEWS
fomp (1.1.0) unstable;

  * Fix phaser plugin data and list phasers in manifest
  * Fix port symbol for MVCLPF4 input gain
    (technically breaks state compatibility, but old symbol was an oversight)
  * Remove invalid reference to old LADSPA pulse_vco binary
  * Add missing default values
  * Add reverb and reverb_amb

 -- David Robillard <d@drobilla.net>  Sat, 25 Jul 2015 13:23:20 -0400

fomp (1.0.0) stable;

  * Initial release

 -- David Robillard <d@drobilla.net>  Sat, 27 Oct 2012 16:54:00 -0400



src/blvco.cc
/*
  Copyright (C) 2003 Fons Adriaensen
    
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#include <stdio.h>
#include <string.h>
#include <math.h>
#include "blvco.h"
#include "exp2ap.h"

extern float exp2ap (float x);


static float _pulse [] = 
{
    0.000000,  0.000366,  0.001458,  0.003117,  0.005028,  0.006756,  0.007806,  0.007704,
    0.006086,  0.002778, -0.002139, -0.008292, -0.015022, -0.021434, -0.026492, -0.029153,
   -0.028511, -0.023957, -0.015297, -0.002869,  0.012417,  0.029105,  0.045302,  0.058853,
    0.067578,  0.069533,  0.063273,  0.048113,  0.024337, -0.006682, -0.042487, -0.079654,
   -0.114036, -0.141088, -0.156243, -0.155364, -0.135210, -0.093843, -0.030932,  0.052092,
    0.152049,  0.264172,  0.382359,  0.499585,  0.608450,  0.701808,  0.773391,  0.818353,
    0.833670,  0.818353,  0.773391,  0.701808,  0.608450,  0.499585,  0.382359,  0.264172,
    0.152049,  0.052092, -0.030932, -0.093843, -0.135210, -0.155364, -0.156243, -0.141088,
   -0.114036, -0.079654, -0.042487, -0.006682,  0.024337,  0.048113,  0.063273,  0.069533,
    0.067578,  0.058853,  0.045302,  0.029105,  0.012417, -0.002869, -0.015297, -0.023957,
   -0.028511, -0.029153, -0.026492, -0.021434, -0.015022, -0.008292, -0.002139,  0.002778,
    0.006086,  0.007704,  0.007806,  0.006756,  0.005028,  0.003117,  0.001458,  0.000366,
    0.000000
};



void Ladspa_VCO_pulse1::setport (PortIndex port, PortData *data)
{
	_port [port] = (float*)data;
}


void Ladspa_VCO_pulse1::active (bool act)
{
    _p = 0.5;
    _w = _y = _z = 0;
    _j = 0;
    memset (_f, 0, (FILLEN + NCOEFF) * sizeof (float));
}


void Ladspa_VCO_pulse1::runproc (SampleCount len, bool add)
{
    int    i, j, n;
    float  *outp, *freq, *expm, *linm;
    float  a, p, r, t, w, dw, y, z, *f;

    outp = _port[OUTP];
    freq = _port[FREQ] - 1;
    expm = _port[EXPM] - 1;
    linm = _port[LINM] - 1;
    
    p = _p;
    w = _w;
    y = _y;
    z = _z;
    j = _j;

    a = 0.2 + 0.8 * _port [FILT][0];
    do
    {
        n = (len > 24) ? 16 : len;
        freq += n;
        expm += n;
        linm += n;
        len -= n;
 
        t = (exp2ap (*freq + _port[OCTN][0] + _port[TUNE][0] + *expm * _port[EXPG][0] + 8.03136)
	     + 1e3 * *linm * _port[LING][0]) / _fsam; 
        if (t < 1e-5) t = 1e-5;
        if (t > 0.5) t = 0.5;
        dw = (t - w) / n;

        while (n--)
	{
	    w += dw; 
            p += w;
            if (p >= 1.0)
	    {   
          	p -= 1.0;
                r = NPHASE * p / w;
                i = (int) r;
                r -= i;
                f = _f + j;
                while (i < NPHASE * NCOEFF)
		{
		   *f++ += r * _pulse [i + 1 ] + (1 - r) * _pulse [i];
                   i += NPHASE;
		}
	    }

            y = _f [j];
            z += a * (y - z);
            *outp++ = z;
            if (++j == FILLEN)
            { 
                j = 0;
		memcpy (_f, _f + FILLEN, NCOEFF * sizeof (float));
                memset (_f + NCOEFF, 0,  FILLEN * sizeof (float));
	    }
	}
    }
    while (len);

    _p = p;
    _w = w;
    _y = y;
    _z = z;
    _j = j;
}



void Ladspa_VCO_saw1::setport (PortIndex port, PortData *data)
{
	_port [port] = (float*)data;
}


void Ladspa_VCO_saw1::active (bool act)
{
    _p = 0.5;
    _w = _x = _y = _z = _d = 0;
    _j = 0;
    memset (_f, 0, (FILLEN + NCOEFF) * sizeof (float));
}


void Ladspa_VCO_saw1::runproc (SampleCount len, bool add)
{
    int    i, j, n;
    float  *outp, *freq, *expm, *linm, *sync;
    float  a, d, p, r, t, w, dw, x, y, z, *f;

    outp = _port[OUTP];
    freq = _port[FREQ] - 1;
    expm = _port[EXPM] - 1;
    linm = _port[LINM] - 1;
    sync = _port[SYNC];

    p = _p;
    w = _w;
    x = _x;
    y = _y;
    z = _z;
    d = _d;
    j = _j;

    a = 0.2 + 0.8 * _port [FILT][0];
    do
    {
        n = (len > 24) ? 16 : len;
        freq += n;
        expm += n;
        linm += n;
        len -= n;
 
        t = (exp2ap (*freq + _port[OCTN][0] + _port[TUNE][0] + *expm * _port[EXPG][0] + 8.03136 + d)
	     + 1e3 * *linm * _port[LING][0]) / _fsam; 
        if (t < 1e-5) t = 1e-5;
        if (t > 0.5) t = 0.5;
        dw = (t - w) / n;

        while (n--)
	{
	    w += dw; 
            p += w;
            if (p >= 1.0)
	    {   
		p -= 1.0;
                r = NPHASE * p / w;
                i = (int) r;
                r -= i;
                f = _f + j;
                while (i < NPHASE * NCOEFF)
		{
		   *f++ += r * _pulse [i + 1 ] + (1 - r) * _pulse [i];
                   i += NPHASE;
		}
	    }
            
            x += _f [j] - w * (0.01 * y + 0.2 * x + 1);
            y += 6.3 * w * x;
            z += a * (x - z);
            *outp++ = z;
            d += 0.01 * (y * *sync++ - d);
            if (++j == FILLEN)
            { 
                j = 0;
		memcpy (_f, _f + FILLEN, NCOEFF * sizeof (float));
                memset (_f + NCOEFF, 0,  FILLEN * sizeof (float));
	    }
	}
    }
    while (len);

    _p = p;
    _w = w;
    _x = x;
    _y = y;
    _z = z;
    _d = d;
    _j = j;
}



void Ladspa_VCO_rec1::setport (PortIndex port, PortData *data)
{
    _port [port] = (float*)data;
}


void Ladspa_VCO_rec1::active (bool act)
{
    _p = 0.0;
    _x = _y = _z = _d = 0;
    _w = 0;
    _b = 0.5;
    _j = _k = 0;
    memset (_f, 0, (FILLEN + NCOEFF) * sizeof (float));
}


void Ladspa_VCO_rec1::runproc (SampleCount len, bool add)
{
    int    i, j, k, n;
    float  *outp, *freq, *expm, *linm, *wavm, *sync;
    float  a, b, db, d, p, r, t, w, dw, x, y, z, *f;

    outp = _port[OUTP];
    freq = _port[FREQ] - 1;
    expm = _port[EXPM] - 1;
    linm = _port[LINM] - 1;
    wavm = _port[WAVM] - 1;
    sync = _port[SYNC];

    p = _p;
    w = _w;
    b = _b;
    x = _x;
    y = _y;
    z = _z;
    d = _d;
    j = _j;
    k = _k;

    a = 0.2 + 0.8 * _port [FILT][0];
    do
    {
        n = (len > 24) ? 16 : len;
        freq += n;
        expm += n;
        linm += n;
        wavm += n;
        len -= n;
 
        t = (exp2ap (*freq + _port[OCTN][0] + _port[TUNE][0] + *expm * _port[EXPG][0] + 8.03136 + d)
	     + 1e3 * *linm * _port[LING][0]) / _fsam; 
        if (t < 1e-5) t = 1e-5;
        if (t > 0.5) t = 0.5;
        dw = (t - w) / n;
        t = 0.5 * (1.0 + _port [WAVE][0] + *wavm * _port [WMOD][0]);         
        if (t < 0.02) t = 0.02;
        if (t > 0.98) t = 0.98;
        db = (t - b) / n;

        while (n--)
	{
	    w += dw; 
            b += db;
            p += w;
            t = k ? 1.0 : b;
            while (p >= t)
	    {
                if (k)
		{
                    k = 0; 
		    p -= 1.0;
                    r = NPHASE * p / w;
                    t = b;
                    i = (int) r;
                    r -= i;
                    f = _f + j;
                    while (i < NPHASE * NCOEFF)
		    {
		        *f++ += r * _pulse [i + 1 ] + (1 - r) * _pulse [i];
                        i += NPHASE;
		    }
  		}
                else
		{ 
                    k = 1;
                    r = NPHASE * (p - t) / w;
                    t = 1.0;
                    i = (int) r;
                    r -= i;
                    f = _f + j;
                    while (i < NPHASE * NCOEFF)
		    {
		        *f++ -= r * _pulse [i + 1 ] + (1 - r) * _pulse [i];
                        i += NPHASE;
		    }
  		}
	    }
            x -= w * (0.01 * y + 0.2 * x);
            x += _f [j];
            y += 6.3 * w * x;
            z += a * (x - z);
            *outp++ = z;
            d += 0.01 * (y * *sync++ - d);
            if (++j == FILLEN)
            { 
                j = 0;
		memcpy (_f, _f + FILLEN, NCOEFF * sizeof (float));
                memset (_f + NCOEFF, 0,  FILLEN * sizeof (float));
	    }
	}
    }
    while (len);

    _p = p;
    _w = w;
    _b = b;
    _x = x;
    _y = y;
    _z = z;
    _d = d;
    _j = j;
    _k = k;
}


src/mvchpf24.cc
/*
  Copyright (C) 2003-2008 Fons Adriaensen <fons@kokkinizita.net>
    
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#include <stdio.h>
#include <math.h>
#include "mvchpf24.h"
#include "exp2ap.h"


void Ladspa_Mvchpf1::setport (PortIndex port, PortData *data)
{
    _port [port] = (float*)data;
}


void Ladspa_Mvchpf1::active (bool act)
{
    _c1 = _c2 = _c3 = _c4 = _x = 0;
    _w = 1e5;
}


void Ladspa_Mvchpf1::runproc (SampleCount len, bool add)
{
    int   k;
    float *p0, *p1, *p2, *p3;
    float c1, c2, c3, c4;
    float g0, g1, d, w, dw, t, x, y;

    p0 = _port [AINP];
    p1 = _port [AOUT];
    p2 = _port [AFREQ] - 1;
    p3 = _port [AFMOD] - 1;
    g0 = exp2ap (0.1661f * _port [CINP][0]) / 4;
    g1 = exp2ap (0.1661f * _port [COUT][0]) * 4;
    if (add) g1 *= _gain;

    c1 = _c1;
    c2 = _c2;
    c3 = _c3;
    c4 = _c4;
    w = _w; 
    x = _x;

    do
    {
        k = (len > 24) ? 16 : len;
        p2 += k;
        p3 += k;
        len -= k;

        t = _fsam / exp2ap (_port [CFMOD][0] * *p3 + *p2 + _port [CFREQ][0] + 9.2f);
        if (t < 2) t = 2;
        dw = (t - w) / k;

        while (k--)
	{
            w += dw;                        

            x = y = *p0++ * g0 - 0.3f * x;

            d = x - c1 + 1e-10f;
            t = d * d;
            d *= (1 + t) / (w + t);            
            c1 += d;
            x -= c1;
            c1 += d;

            d = x - c2 + 1e-10f;
            t = d * d;
            d *= (1 + t) / (w + t);            
            c2 += d;
            x -= c2;
            c2 += d;

            d = x - c3 + 1e-10f;
            t = d * d;
            d *= (1 + t) / (w + t);            
            c3 += d;
            x -= c3;
            c3 += d;

            d = x - c4 + 1e-10f;
            t = d * d;
            d *= (1 + t) / (w + t);            
            c4 += d;
            x -= c4;
            c4 += d;

            if (add) *p1++ += g1 * x;
            else     *p1++  = g1 * x;

            x -= y;
	}
    }
    while (len);

    _c1 = c1;
    _c2 = c2;
    _c3 = c3;
    _c4 = c4;
    _w = w;
    _x = x;
}





src/ladspaplugin.h
/*
    Copyright (C) 2003-2008 Fons Adriaensen <fons@kokkinizita.net>
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#ifndef __LADSPAPLUGIN_H
#define __LADSPAPLUGIN_H

#include <lv2/lv2plug.in/ns/lv2core/lv2.h>


/* This class name is now a misnomer but I have left it as-is to minimize
   fragmentation.  The following typedefs serve the same purpose, so the
   plugins can be built with the correct LV2 type (e.g. uint32_t instead of
   unsigned long) but the actual DSP code has only been changed to use the
   typedef rather than 'hard' switching it to the new type.  This way, it
   should be straightforward to make the same code build as both LADSPA and LV2
   plugins, though since the upstream situation is unclear I have not done
   done so.  ~ David Robillard, Aug. 2012 */

typedef unsigned long SampleRate;
typedef unsigned long SampleCount;
typedef unsigned long PortIndex;
typedef void          PortData;

class LadspaPlugin
{
public:

  LadspaPlugin (SampleRate fsam) : _gain (1.0), _fsam (fsam) {}

  virtual void setport (PortIndex port, PortData *data) = 0;  
  virtual void active  (bool act) = 0;  
  virtual void runproc (SampleCount len, bool add) = 0;  
  virtual ~LadspaPlugin (void) {}

  void setgain (float gain) { _gain = gain; }

protected:

  float _gain;
  float _fsam;
};



#endif


src/mvclpf24_lv2.cc
/*
  Copyright (C) 2012 David Robillard <d@drobilla.net>
    
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stddef.h>

#include "mvclpf24.h"

extern "C" {

static void
connect_port(LV2_Handle instance,
             uint32_t   port,
             void*      data)
{
	((LadspaPlugin*)instance)->setport(port, data);
}

static void
activate(LV2_Handle instance)
{
    ((LadspaPlugin*)instance)->active(true);
}

static void
run(LV2_Handle instance, uint32_t n_samples)
{
    ((LadspaPlugin*)instance)->runproc(n_samples, false);
}

static void
deactivate(LV2_Handle instance)
{
    ((LadspaPlugin*)instance)->active(false);
}

static void
cleanup(LV2_Handle instance)
{
	delete ((LadspaPlugin*)instance);
}

static LV2_Handle
instantiate1(const LV2_Descriptor*     descriptor,
             double                    rate,
             const char*               bundle_path,
             const LV2_Feature* const* features)
{
	return new Ladspa_Moogvcf1(rate);
}

static LV2_Handle
instantiate2(const LV2_Descriptor*     descriptor,
             double                    rate,
             const char*               bundle_path,
             const LV2_Feature* const* features)
{
	return new Ladspa_Moogvcf1(rate);
}
static LV2_Handle
instantiate3(const LV2_Descriptor*     descriptor,
             double                    rate,
             const char*               bundle_path,
             const LV2_Feature* const* features)
{
	return new Ladspa_Moogvcf1(rate);
}
static LV2_Handle
instantiate4(const LV2_Descriptor*     descriptor,
             double                    rate,
             const char*               bundle_path,
             const LV2_Feature* const* features)
{
	return new Ladspa_Moogvcf1(rate);
}

static const LV2_Descriptor descriptors[4] = {
	{ "http://drobilla.net/plugins/fomp/mvclpf1",
	  instantiate1,
	  connect_port,
	  activate,
	  run,
	  deactivate,
	  cleanup,
	  NULL },
	{ "http://drobilla.net/plugins/fomp/mvclpf2",
	  instantiate2,
	  connect_port,
	  activate,
	  run,
	  deactivate,
	  cleanup,
	  NULL },
	{ "http://drobilla.net/plugins/fomp/mvclpf3",
	  instantiate3,
	  connect_port,
	  activate,
	  run,
	  deactivate,
	  cleanup,
	  NULL },
	{ "http://drobilla.net/plugins/fomp/mvclpf4",
	  instantiate4,
	  connect_port,
	  activate,
	  run,
	  deactivate,
	  cleanup,
	  NULL }
};

LV2_SYMBOL_EXPORT
const LV2_Descriptor*
lv2_descriptor(uint32_t index)
{
	if (index < 4) {
		return &descriptors[index];
	}
	return NULL;
}

}  // extern "C"


src/mvclpf24.cc
/*
  Copyright (C) 2003-2008 Fons Adriaensen <fons@kokkinizita.net>
    
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#include <stdio.h>
#include <math.h>
#include "mvclpf24.h"
#include "exp2ap.h"


void Ladspa_Moogvcf1::setport (PortIndex port, PortData *data)
{
	_port [port] = (float*)data;
}


void Ladspa_Moogvcf1::active (bool act)
{
    _c1 = _c2 = _c3 = _c4 = _c5 = _w = _r = 0;
}


void Ladspa_Moogvcf1::runproc (SampleCount len, bool add)
{
    int   k;
    float *p0, *p1, *p2, *p3, *p4;
    float c1, c2, c3, c4, c5;
    float g0, g1, r, dr, w, dw, x, t;

    p0 = _port [A_INPUT];
    p1 = _port [A_OUTPUT];
    p2 = _port [A_FREQ] - 1;
    p3 = _port [A_FMOD] - 1;
    p4 = _port [A_RESO] - 1;
    g0 = exp2ap (0.1661f * _port [C_IPGAIN][0]) / 4;
    g1 = exp2ap (0.1661f * _port [C_OPGAIN][0]) * 4;
    if (add) g1 *= _gain;

    c1 = _c1 + 1e-6f;
    c2 = _c2;
    c3 = _c3;
    c4 = _c4;
    c5 = _c5;
    w = _w; 
    r = _r;
 
    do
    {
        k = (len > 24) ? 16 : len;
        p2 += k;
        p3 += k;
        p4 += k;
        len -= k;

        t = exp2ap (_port [C_FMODG][0] * *p3 + _port [C_FREQ][0] + *p2 + 10.82f) / _fsam;
	if (t < 0.8f) t *= 1 - 0.4f * t - 0.125f * t * t;
        else 
	{
            t *= 0.6f; 
            if (t > 0.92f) t = 0.92f;
	}
        dw = (t - w) / k;

        t = _port [C_RESOG][0] * *p4 + _port [C_RESO][0];
        if (r > 1) r = 1;
        if (r < 0) r = 0;
        dr = (t - r) / k;  

        while (k--)
	{
            w += dw;                        
            r += dr;
	    x = -4.2 * r * c5 + *p0++ * g0 + 1e-10f;
            t = c1 / (1 + fabsf (c1));
            c1 += w * (x - t);
            x = c1 / (1 + fabsf (c1));
            c2 += w * (x  - c2);
            c3 += w * (c2 - c3);
            c4 += w * (c3 - c4);
	    if (add) *p1++ += g1 * c4;
	    else     *p1++  = g1 * c4;
	    c5 += 0.5f * (c4 - c5);
	}
    }
    while (len);

    _c1 = c1;
    _c2 = c2;
    _c3 = c3;
    _c4 = c4;
    _c5 = c5;
    _w = w;
    _r = r;
}



void Ladspa_Moogvcf2::setport (PortIndex port, PortData *data)
{
    _port [port] = (float*)data;
}


void Ladspa_Moogvcf2::active (bool act)
{
    _c1 = _c2 = _c3 = _c4 = _c5 = _w = _r = 0;
}


void Ladspa_Moogvcf2::runproc (SampleCount len, bool add)
{
    int   k;
    float *p0, *p1, *p2, *p3, *p4;
    float c1, c2, c3, c4, c5;
    float g0, g1, r, dr, w, dw, x, t;

    p0 = _port [A_INPUT];
    p1 = _port [A_OUTPUT];
    p2 = _port [A_FREQ] - 1;
    p3 = _port [A_FMOD] - 1;
    p4 = _port [A_RESO] - 1;
    g0 = exp2ap (0.1661f * _port [C_IPGAIN][0]) / 4;
    g1 = exp2ap (0.1661f * _port [C_OPGAIN][0]) * 4;
    if (add) g1 *= _gain;

    c1 = _c1 + 1e-6f;
    c2 = _c2;
    c3 = _c3;
    c4 = _c4;
    c5 = _c5;
    w = _w; 
    r = _r;

    do
    {
        k = (len > 24) ? 16 : len;
        p2 += k;
        p3 += k;
        p4 += k;
        len -= k;

        t = exp2ap (_port [C_FMODG][0] * *p3 + _port [C_FREQ][0] + *p2 + 10.71f) / _fsam;
	if (t < 0.8f) t *= 1 - 0.4f * t - 0.125f * t * t;
        else 
	{
            t *= 0.6f; 
            if (t > 0.92f) t = 0.92f;
	}
        dw = (t - w) / k;

        t = _port [C_RESOG][0] * *p4 + _port [C_RESO][0];
        if (t > 1) t = 1;
        if (t < 0) t = 0;
        dr = (t - r) / k;  

        while (k--)
	{
            w += dw;                        
            r += dr;

	    x = -4.5f * r * c5 + *p0++ * g0 + 1e-10f;
//	    x = tanhf (x); 
            x /= sqrtf (1 + x * x);
            c1 += w * (x  - c1) / (1 + c1 * c1);            
            c2 += w * (c1 - c2) / (1 + c2 * c2);            
            c3 += w * (c2 - c3) / (1 + c3 * c3);            
            c4 += w * (c3 - c4) / (1 + c4 * c4);            

	    if (add) *p1++ += g1 * (c4);
	    else     *p1++  = g1 * (c4);
	    c5 += 0.5f * (c4 - c5);
	}
    }
    while (len);

    _c1 = c1;
    _c2 = c2;
    _c3 = c3;
    _c4 = c4;
    _c5 = c5;
    _w = w;
    _r = r;
}



void Ladspa_Moogvcf3::setport (PortIndex port, PortData *data)
{
    _port [port] = (float*)data;
}


void Ladspa_Moogvcf3::active (bool act)
{
    _c1 = _c2 = _c3 = _c4 = _c5 = _w = _r = 0;
}


void Ladspa_Moogvcf3::runproc (SampleCount len, bool add)
{
    int   k;
    float *p0, *p1, *p2, *p3, *p4;
    float c1, c2, c3, c4, c5;
    float g0, g1, r, dr, w, dw, x, t, d;

    p0 = _port [A_INPUT];
    p1 = _port [A_OUTPUT];
    p2 = _port [A_FREQ] - 1;
    p3 = _port [A_FMOD] - 1;
    p4 = _port [A_RESO] - 1;
    g0 = exp2ap (0.1661f * _port [C_IPGAIN][0]) / 4;
    g1 = exp2ap (0.1661f * _port [C_OPGAIN][0]) * 4;
    if (add) g1 *= _gain;

    c1 = _c1;
    c2 = _c2;
    c3 = _c3;
    c4 = _c4;
    c5 = _c5;
    w = _w; 
    r = _r;

    do
    {
        k = (len > 24) ? 16 : len;
        p2 += k;
        p3 += k;
        p4 += k;
        len -= k;

        t = exp2ap (_port [C_FMODG][0] * *p3 + _port [C_FREQ][0] + *p2 + 9.70f) / _fsam;
        if (t < 0.75f) t *= 1.005f - t * (0.624f - t * (0.65f - t * 0.54f));
        else
	{
	    t *= 0.6748f;
            if (t > 0.82f) t = 0.82f;
	}
        dw = (t - w) / k;

        t = _port [C_RESOG][0] * *p4 + _port [C_RESO][0];
        if (t > 1) t = 1;
        if (t < 0) t = 0;
        dr = (t - r) / k;  

        while (k--)
	{
            w += dw;                        
            r += dr;

	    x = *p0 * g0 - (4.3f - 0.2f * w) * r * c5 + 1e-10f; 
//            x = tanhf (x);
            x /= sqrtf (1 + x * x);
            d = w * (x  - c1) / (1 + c1 * c1);            
            x = c1 + 0.77f * d;
            c1 = x + 0.23f * d;        
            d = w * (x  - c2) / (1 + c2 * c2);            
            x = c2 + 0.77f * d;
            c2 = x + 0.23f * d;        
            d = w * (x  - c3) / (1 + c3 * c3);            
            x = c3 + 0.77f * d;
            c3 = x + 0.23f * d;        
            d = w * (x  - c4);
            x = c4 + 0.77f * d;
            c4 = x + 0.23f * d;        
            c5 += 0.85f * (c4 - c5);

	    x = *p0++ * g0 -(4.3f - 0.2f * w) * r * c5;
//            x = tanhf (x);
            x /= sqrtf (1 + x * x);
            d = w * (x  - c1) / (1 + c1 * c1);            
            x = c1 + 0.77f * d;
            c1 = x + 0.23f * d;        
            d = w * (x  - c2) / (1 + c2 * c2);            
            x = c2 + 0.77f * d;
            c2 = x + 0.23f * d;        
            d = w * (x  - c3) / (1 + c3 * c3);            
            x = c3 + 0.77f * d;
            c3 = x + 0.23f * d;        
            d = w * (x  - c4);
            x = c4 + 0.77f * d;
            c4 = x + 0.23f * d;        
            c5 += 0.85f * (c4 - c5);

	    if (add) *p1++ += g1 * c4;
            else     *p1++  = g1 * c4;
	}
    }
    while (len);

    _c1 = c1;
    _c2 = c2;
    _c3 = c3;
    _c4 = c4;
    _c5 = c5;
    _w = w;
    _r = r;
}


void Ladspa_Moogvcf4::setport (PortIndex port, PortData *data)
{
    _port [port] = (float*)data;
}


void Ladspa_Moogvcf4::active (bool act)
{
    _c1 = _c2 = _c3 = _c4 = _c5 = _w = _r = 0;
}


void Ladspa_Moogvcf4::runproc (SampleCount len, bool add)
{
    int   k, op;
    float *p0, *p1, *p2, *p3, *p4;
    float c1, c2, c3, c4, c5;
    float g0, g1, r, dr, w, dw, x, t, d, y;

    p0 = _port [A_INPUT];
    p1 = _port [A_OUTPUT];
    p2 = _port [A_FREQ] - 1;
    p3 = _port [A_FMOD] - 1;
    p4 = _port [A_RESO] - 1;
    g0 = exp2ap (0.1661f * _port [C_IPGAIN][0]) / 4;
    g1 = exp2ap (0.1661f * _port [C_OPGAIN][0]) * 4;
    op = (int)(floorf (_port [C_LPFILT][0] + 0.5));
    if (add) g1 *= _gain;

    c1 = _c1 + 1e-6f;
    c2 = _c2;
    c3 = _c3;
    c4 = _c4;
    c5 = _c5;
    w = _w; 
    r = _r;

    do
    {
        k = (len > 24) ? 16 : len;
        p2 += k;
        p3 += k;
        p4 += k;
        len -= k;

        t = exp2ap (_port [C_FMODG][0] * *p3 + _port [C_FREQ][0] + *p2 + 9.70f) / _fsam;
        if (t < 0.75f) t *= 1.005f - t * (0.624f - t * (0.65f - t * 0.54f));
        else
	{
	    t *= 0.6748f;
            if (t > 0.82f) t = 0.82f;
	}
        dw = (t - w) / k;

        t = _port [C_RESOG][0] * *p4 + _port [C_RESO][0];
        if (t > 1) t = 1;
        if (t < 0) t = 0;
        dr = (t - r) / k;  

        while (k--)
	{
            w += dw;                        
            r += dr;

	    x = *p0 * g0 - (4.3f - 0.2f * w) * r * c5 + 1e-10f; 
//            x = tanhf (x);
            x /= sqrtf (1 + x * x);
	    d = w * (x  - c1) / (1 + c1 * c1);           
            x = c1 + 0.77f * d;
            c1 = x + 0.23f * d;        
            d = w * (x  - c2) / (1 + c2 * c2);            
            x = c2 + 0.77f * d;
            c2 = x + 0.23f * d;        
            d = w * (x  - c3) / (1 + c3 * c3);            
            x = c3 + 0.77f * d;
            c3 = x + 0.23f * d;        
            d = w * (x  - c4);
            x = c4 + 0.77f * d;
            c4 = x + 0.23f * d;        
            c5 += 0.85f * (c4 - c5);

	    x = y = *p0++ * g0 -(4.3f - 0.2f * w) * r * c5;
//            x = tanhf (x);
            x /= sqrtf (1 + x * x);
            d = w * (x  - c1) / (1 + c1 * c1);            
            x = c1 + 0.77f * d;
            c1 = x + 0.23f * d;        
            d = w * (x  - c2) / (1 + c2 * c2);            
            x = c2 + 0.77f * d;
            c2 = x + 0.23f * d;        
            d = w * (x  - c3) / (1 + c3 * c3);            
            x = c3 + 0.77f * d;
            c3 = x + 0.23f * d;        
            d = w * (x  - c4);
            x = c4 + 0.77f * d;
            c4 = x + 0.23f * d;        
            c5 += 0.85f * (c4 - c5);

            switch (op)
	    {
            case 1: y = c1; break;
            case 2: y = c2; break;
            case 3: y = c3; break;
            case 4: y = c4; break;
	    }

	    if (add) *p1++ += g1 * y;
            else     *p1++  = g1 * y;
	}
    }
    while (len);

    _c1 = c1;
    _c2 = c2;
    _c3 = c3;
    _c4 = c4;
    _c5 = c5;
    _w = w;
    _r = r;
}


